pkg/controller/cloud/node_gc_controller.go

```diff
@@ -0,0 +1,215 @@
+			continue
+		}
+		// If the known node status says that Node is NotReady, then check if the node has been removed
+		// from the cloud provider. If node cannot be found in cloudprovider, then delete the node immediately
+		if currentReadyCondition != nil {
+			if currentReadyCondition.Status != v1.ConditionTrue {
+				// we need to check this first to get taint working in similar in all cloudproviders
+				// current problem is that shutdown nodes are not working in similar way ie. all cloudproviders
+				// does not delete node from kubernetes cluster when instance it is shutdown see issue #46442
+				shutdown, err := shutdownInCloudProvider(context.TODO(), c.cloud, node)
+				if err != nil {
+					glog.Errorf("Error checking if node %s is shutdown: %v", node.Name, err)
+				}
+
+				if shutdown && err == nil {
+					// if node is shutdown add shutdown taint
+					err = controller.AddOrUpdateTaintOnNode(c.kubeClient, node.Name, ShutdownTaint)
+					if err != nil {
+						glog.Errorf("Error patching node taints: %v", err)
+					}
+					// Continue checking the remaining nodes since the current one is shutdown.
+					continue
+				}
+
+				// Check with the cloud provider to see if the node still exists. If it
+				// doesn't, delete the node immediately.
+				exists, err := ensureNodeExistsByProviderID(instances, node)
+				if err != nil {
+					glog.Errorf("Error checking if node %s exists: %v", node.Name, err)
+					continue
+				}
+
+				if exists {
+					// Continue checking the remaining nodes since the current one is fine.
+					continue
+				}
+
+				glog.V(2).Infof("Deleting node since it is no longer present in cloud provider: %s", node.Name)
+
+				ref := &v1.ObjectReference{
+					Kind:      "Node",
+					Name:      node.Name,
+					UID:       types.UID(node.UID),
+					Namespace: "",
+				}
+				glog.V(2).Infof("Recording %s event message for node %s", "DeletingNode", node.Name)
+
+				c.recorder.Eventf(ref, v1.EventTypeNormal, fmt.Sprintf("Deleting Node %v because it's not present according to cloud provider", node.Name), "Node %s event: %s", node.Name, "DeletingNode")
+
+				go func(nodeName string) {
+					defer utilruntime.HandleCrash()
+					if err := c.kubeClient.CoreV1().Nodes().Delete(nodeName, nil); err != nil {
+						glog.Errorf("unable to delete node %q: %v", nodeName, err)
+					}
+				}(node.Name)
+
+			} else {
+				// if taint exist remove taint
+				err = controller.RemoveTaintOffNode(c.kubeClient, node.Name, node, ShutdownTaint)
+				if err != nil {
+					glog.Errorf("Error patching node taints: %v", err)
+				}
+			}
+		}
+	}
+}
+
+// shutdownInCloudProvider returns true if the node is shutdowned in
+// cloud provider.
+func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node) (bool, error) {
+	instances, ok := cloud.Instances()
+	if !ok {
+		return false, errors.New("cloud provider does not support instances")
```

Comment:
Wouldn't it make more sense to check this in start/run and just skip the controller entirely in that case?

Reply:
I had the same thought and added it here https://github.com/kubernetes/kubernetes/pull/70344/files#diff-da5354d582a4b27381b8cee161e46163R82, but then figured it can't hurt to check it here too

Reply:
I did what service controller does, which is to log the error but still return nil so kube-controller-manager doesn't crash. See [here](https://github.com/kubernetes/kubernetes/pull/70344/files#diff-0ca8b7870d367d5fc309a2e737a6eed2R159)
```go
func startCloudNodeGarbageCollectorController(ctx ControllerContext) (http.Handler, bool, error) {
	cloudNodeGCController, err := cloudcontroller.NewCloudNodeGarbageCollectorController(
		ctx.ClientBuilder.ClientOrDie("cloud-node-gc-controller"),
		ctx.Cloud,
		ctx.ComponentConfig.KubeCloudShared.NodeMonitorPeriod.Duration,
	)
	if err != nil {
		// the controller manager should continue to run if the "Instances" interface is not
		// supported, though it's unlikely for a cloud provider to not support it
		glog.Errorf("failed to start cloud node garbage collector controller: %v", err)
		return nil, false, nil
	}
	go cloudNodeGCController.Run(ctx.Stop)
	return nil, true, nil
}
```

Reply:
Good call